firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp

```diff
@@ -0,0 +1,391 @@
+      output_payload.copy_from(
+        input_payload.buffer(), input_payload.size(), TestCRCElementHeaderProps::payload_offset);
+
+      auto computed_crc = crc_element.compute_body_crc(output_payload, crc32c);
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element is initalised with payload 0x00") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    input_payload.push_back(0x00);
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("the crc is computed from the payload") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0x527d5351;
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      output_payload.resize(TestCRCElementHeaderProps::header_size + input_payload.size());
+      output_payload.copy_from(
+        input_payload.buffer(), input_payload.size(), TestCRCElementHeaderProps::payload_offset);
+
+      auto computed_crc = crc_element.compute_body_crc(output_payload, crc32c);
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element is initalised with payload 0x01") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    input_payload.push_back(0x01);
+
+    TestCRCElement crc_element{input_payload};
+
+    WHEN("the crc is computed from the payload") {
+      // Calculated using the Sunshine Online CRC Calculator
+      uint32_t expected_crc = 0xa016d052;
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      output_payload.resize(TestCRCElementHeaderProps::header_size + input_payload.size());
+      output_payload.copy_from(
+        input_payload.buffer(), input_payload.size(), TestCRCElementHeaderProps::payload_offset);
+
+      auto computed_crc = crc_element.compute_body_crc(output_payload, crc32c);
+
+      THEN("the computed crc is equal to the expected crc") {
+        REQUIRE(computed_crc == expected_crc);
+      }
+    }
+  }
+
+  GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]") {
+    TestCRCElementProps::PayloadBuffer input_payload;
+    auto data = PF::Util::make_array<uint8_t>(0x01, 0x02, 0x05);
+    for (auto& bytes : data) {
+      input_payload.push_back(bytes);
+    }
+
+    TestCRCElement crc_element{input_payload};
+    WHEN("given payload data is written to it") {
+
+      TestCRCElementProps::PayloadBuffer output_payload;
+
+      auto write_status = crc_element.write(output_payload, crc32c);
+      auto crc = crc_element.crc();
+      auto crc_compute = crc32c.compute(input_payload.buffer(), input_payload.size());
+
+      THEN("crc computed in write method is same as crcchecker compute") {
+        REQUIRE(write_status == PF::IndexStatus::ok);
+        REQUIRE(crc == crc_compute);
+      }
+    }
+  }
+
+  GIVEN("crc element is initalised with 0 bytes") {
```

Comment:
This Given looks like an outdated copy of the first Given, do we still need it?

Reply:
After latest changes, this will come under scenario for the write function therefore a different given ( where a crc element is initalised with 0 bytes)